File: src/main/java/io/protobase/f7/formulas/info/ERRORTYPE.java
package io.protobase.f7.formulas.info;

import io.protobase.f7.errors.F7Exception;
import io.protobase.f7.errors.NAException;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class ERRORTYPE extends AbstractFormula {
  public static FormulaName NAME = FormulaName.ERRORTYPE;
  public static ERRORTYPE SELF = new ERRORTYPE();

  public ERRORTYPE() {
    super();
  }

  public ERRORTYPE(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    try {
      Object first = Converters.first(collateralLookup.apply(origin, values[0]));
      if (first instanceof F7Exception) {
        return (double) Converters.castAsF7Error(first).getName().getCode();
      }
      return new NAException("Formula ERRORTYPE parameter 1 is not an error.");
    } catch (F7Exception exception) {
      return (double) exception.getName().getCode();
    }
  }
}

File: src/main/js/formulas/info/ERRORTYPE.ts
import { F7Exception } from "../../errors/F7Exception";
import { getExceptionCode } from "../../errors/F7ExceptionName";
import { NAException } from "../../errors/NAException";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class ERRORTYPE extends AbstractFormula {
  static SELF: ERRORTYPE = new ERRORTYPE();
  NAME = FormulaName.ERRORTYPE;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    try {
      const first = Converters.first(this.collateralLookup(origin, values[0]));
      if (first instanceof F7Exception) {
        return getExceptionCode(Converters.castAsF7Exception(first).name);
      }
      return new NAException("Formula ERRORTYPE parameter 1 is not an error.");
    } catch (error) {
      return getExceptionCode(Converters.castAsF7Exception(error as F7Exception).name);
    }
  }
}
